File: mod/Sitrep.Contract/VesselParts.cs
using System.Collections.Generic;
#if SITREP_CODEGEN
using Reinforced.Typings.Attributes;
#endif

namespace Sitrep.Contract;

/// <summary>
/// The <c>vessel.parts</c> channel payload: the active vessel's full
/// part-tree topology (P1b slice 2), the foundation ShipMap / PowerSystems
/// topology / ThermalStatus all build on. A SINGLE WRAPPER OBJECT (or
/// <c>null</c> when there is no active vessel / no topology group this tick),
/// so the Topic tag sits on this type directly with the default
/// <c>IsArray = false</c>: same posture as <see cref="VesselStructure"/> and
/// the sibling structured <c>vessel.*</c> channels, NOT the bare-array
/// <c>parts.robotics</c>.
///
/// <para><b>Thermal folds in here.</b> Per-part temperatures ride each
/// <see cref="VesselPart"/> (<see cref="VesselPart.CurrentTemp"/>/
/// <see cref="VesselPart.MaxTemp"/>/<see cref="VesselPart.SkinTemp"/>/
/// <see cref="VesselPart.SkinMaxTemp"/>), so the hottest-part / engine /
/// heat-shield rollups are SDK-DERIVABLE on top of this channel, there is no
/// separate <c>therm.hottestPart*</c> Topic (v-topology-redesign.md). The
/// existing <c>vessel.thermal</c> rollup channel is NOT removed by this build;
/// that is a later cleanup.</para>
///
/// <para><b>Typing-only mirror.</b> This reproduces, field-for-field, the
/// exact serialized shape <c>Sitrep.Host.VesselPartsViewProvider.ToWire</c>
/// already emits (same names, same camelCase wire keys via
/// <c>RtConfig.CamelCaseForProperties</c>, same units). It is NOT serialized
/// itself: the wire is written by <c>JsonWriter</c> walking the provider's
/// dictionary: so adding it changes no bytes.</para>
/// </summary>
[SitrepContract]
#if SITREP_CODEGEN
[TsInterface]
#endif
[SitrepTopic("vessel.parts")]
public class VesselParts
{
    /// <summary>Every part on the active vessel this tick, in vessel part-list order. Always present (possibly empty); a vessel-less tick yields a <c>null</c> payload, not an empty list.</summary>
    public List<VesselPart> Parts { get; set; } = new();

    public PayloadMeta Meta { get; set; } = new();
}

/// <summary>
/// One part in the <see cref="VesselParts.Parts"/> tree. Provenance-scoped
/// like <see cref="VesselStructure"/> (whole payload absent when there is no
/// vessel), so the always-present required fields (<see cref="Id"/>/
/// <see cref="Name"/>/<see cref="Position"/>/<see cref="DryMass"/>/
/// <see cref="InverseStage"/>/<see cref="MaxTemp"/>) are non-nullable, while
/// the genuinely-optional ones (<see cref="ParentId"/> null for the root,
/// <see cref="Up"/>, <see cref="SkinMaxTemp"/>/<see cref="CurrentTemp"/>/
/// <see cref="SkinTemp"/> unset before physics runs,
/// <see cref="FuelLineTargetId"/>) are nullable.
///
/// <para><b>Join key.</b> <see cref="Id"/> is <c>Part.flightID</c> stringified,
/// the SAME string form <c>parts.power</c>/<c>parts.robotics</c>'s
/// <c>partId</c> uses, so a consumer (RoboticsConsole, PowerSystems) can
/// id-join a part across those channels. <see cref="ParentId"/> and
/// <see cref="FuelLineTargetId"/> are the same string form for the same
/// reason. flightID's stability across a docking/undocking round-trip is a
/// KSP-side caveat carried forward from the design's open questions.</para>
/// </summary>
[SitrepContract]
#if SITREP_CODEGEN
[TsInterface]
#endif
public class VesselPart
{
    /// <summary><c>Part.flightID</c> stringified: the tree/cross-channel join key. Empty string only for the uninitialized-0 sentinel (no live flight id yet).</summary>
    [SitrepUnit(Units.Id)]
    public string Id { get; set; } = "";

    /// <summary><c>Part.parent?.flightID</c> stringified; <c>null</c> for the root part.</summary>
    [SitrepUnit(Units.Id)]
    public string? ParentId { get; set; }

    /// <summary><c>Part.partInfo.name</c> (the <c>AvailablePart.name</c> config id, e.g. <c>"solarPanels1"</c>).</summary>
    [SitrepUnit(Units.Text)]
    public string Name { get; set; } = "";

    /// <summary><c>Part.partInfo.title</c> (the display title, e.g. <c>"OX-STAT Photovoltaic Panels"</c>).</summary>
    [SitrepUnit(Units.Text)]
    public string Title { get; set; } = "";

    /// <summary><c>Part.orgPos</c>: the part's original vessel-local position (metres, vessel frame).</summary>
    [SitrepUnit(Units.Metres)]
    [SitrepFrame(Frames.VesselLocal)]
    public Vec3 Position { get; set; } = new();

    /// <summary>The part's local up axis (<c>Part.orgRot * Vector3.up</c>), for orienting flow/thrust glyphs. <c>null</c> on a snapshot recorded before this field existed.</summary>
    [SitrepUnit(Units.Dimensionless)]
    // orgRot is the part's rotation within the construction frame, so the rotated axis lands in
    // the vessel's frame and not the part's own.
    [SitrepFrame(Frames.VesselLocal)]
    public Vec3? Up { get; set; }

    public PartBounds Bounds { get; set; } = new();

    /// <summary><c>Part.mass</c>: dry mass (tonnes).</summary>
    [SitrepUnit(Units.Tonnes)]
    public double DryMass { get; set; }

    /// <summary><c>Part.inverseStage</c> (KSP's own inverted staging numbering, carried forward unchanged; see <see cref="VesselStructure.CurrentStage"/>).</summary>
    [SitrepUnit(Units.Id)]
    public int InverseStage { get; set; }

    /// <summary><c>Part.maxTemp</c>: internal max temperature (K).</summary>
    [SitrepUnit(Units.Kelvin)]
    public double MaxTemp { get; set; }

    /// <summary><c>Part.skinMaxTemp</c> (K); <c>null</c> for the <c>-1</c> "no skin-thermal model" sentinel.</summary>
    [SitrepUnit(Units.Kelvin)]
    public double? SkinMaxTemp { get; set; }

    /// <summary><c>Part.temperature</c>: current internal temperature (K); <c>null</c> for the <c>-1</c> "not yet simulated" sentinel.</summary>
    [SitrepUnit(Units.Kelvin)]
    public double? CurrentTemp { get; set; }

    /// <summary><c>Part.skinTemperature</c>: current skin temperature (K).</summary>
    [SitrepUnit(Units.Kelvin)]
    public double? SkinTemp { get; set; }

    /// <summary><c>Part.partInfo.category</c> (<c>PartCategories</c> enum name, e.g. <c>"Engine"</c>). A display label; see <see cref="CategoryOrdinal"/>.</summary>
    [SitrepUnit(Units.Text)]
    public string Category { get; set; } = "";

    /// <summary>
    /// <see cref="Category"/>'s KSP ORDINAL, typed to
    /// <see cref="KspPartCategory"/>.
    ///
    /// <para>ShipMap picks a part's diagram glyph from this. It used to switch
    /// on the NAME, so a member KSP renamed dropped every part of that category
    /// through to the name/title heuristic underneath: engines drawn as
    /// whatever "engine" happened to match in a part's title, and nothing to
    /// say it had happened.</para>
    ///
    /// <para><c>null</c> when the part had no <c>partInfo</c> to read, the same
    /// case that already leaves <see cref="Category"/> empty. Note
    /// <c>PartCategories.none</c> is <c>-1</c> and is a real value, NOT an
    /// absence.</para>
    /// </summary>
    [SitrepUnit(Units.Enumeration)]
    public KspPartCategory? CategoryOrdinal { get; set; }

    /// <summary>Each <c>PartModule</c>'s CLR class name (e.g. <c>"ModuleEngines"</c>, <c>"CModuleFuelLine"</c>), which is what a client classifies a part by.</summary>
    [SitrepUnit(Units.Text)]
    public List<string> Modules { get; set; } = new();

    /// <summary><c>Part.isRobotic()</c>: a Breaking Ground robotic servo part.</summary>
    [SitrepUnit(Units.Flag)]
    public bool IsRobotics { get; set; }

    /// <summary>True when the part carries a solar panel, alternator, EC-producing converter, or an ElectricCharge resource capacity.</summary>
    [SitrepUnit(Units.Flag)]
    public bool IsPowerRelated { get; set; }

    /// <summary>For a fuel-line part, the stringified <c>flightID</c> of the part it feeds; <c>null</c> otherwise.</summary>
    [SitrepUnit(Units.Id)]
    public string? FuelLineTargetId { get; set; }

    /// <summary>
    /// Every resource this part carries (join key: resource name, e.g.
    /// <c>"ElectricCharge"</c>), storage plus live production/consumption
    /// flow: the per-part live-data slice a client used to have to fetch off
    /// the legacy <c>r.resourceFor[flightId]</c> key.
    /// Empty dict when the part carries no resources.
    /// </summary>
    public Dictionary<string, PartResourceFlow> Resources { get; set; } = new();

    /// <summary>
    /// Per-module behavioural state (solar deployed, engine firing,
    /// parachute armed, etc.): one entry per module on the part that maps
    /// to <see cref="PartModuleState"/>'s vocabulary, in <c>Part.Modules</c>
    /// order. The per-part live-data slice the SDK used to fetch off the
    /// legacy <c>v.partState[flightId]</c> key. Empty list when the part
    /// carries no module of a mapped type.
    /// </summary>
    public List<PartModuleState> ModuleStates { get; set; } = new();

    /// <summary>
    /// Action-group bindings on this part: one entry per bound part action
    /// (<see cref="ActionBinding.Action"/> = <c>BaseAction.guiName</c>, and the
    /// named groups its <c>BaseAction.actionGroup</c> Flags bitmask decodes to).
    /// Per-ACTION, not per-part. Empty when no action on the part is bound to
    /// any group. Retires the legacy <c>f.ag.bindings</c> shim: the client
    /// derives the human-readable action-group caption from this field.
    /// </summary>
    public List<ActionBinding> ActionBindings { get; set; } = new();
}

/// <summary>
/// One action-group binding in <see cref="VesselPart.ActionBindings"/>: a
/// single part action and the named action groups it fires with. <see cref="Groups"/>
/// are the <c>KSPActionGroup</c> enum member names (<c>SAS</c>/<c>RCS</c>/
/// <c>Brakes</c>/<c>Gear</c>/<c>Light</c>/<c>Abort</c>/<c>Stage</c>/
/// <c>Custom01</c>...) the action's Flags bitmask decodes to (<c>None</c> excluded).
/// </summary>
[SitrepContract]
#if SITREP_CODEGEN
[TsInterface]
#endif
public class ActionBinding
{
    /// <summary>The action's PAW label: <c>BaseAction.guiName</c> (e.g. "Toggle", "Extend Panel").</summary>
    [SitrepUnit(Units.Text)]
    public string Action { get; set; } = "";

    /// <summary>Named KSPActionGroup groups this action is bound to (e.g. <c>["SAS","Custom01"]</c>). Never empty, an action bound to no group isn't emitted. Display labels; see <see cref="GroupsMask"/>.</summary>
    [SitrepUnit(Units.Text)]
    public List<string> Groups { get; set; } = new();

    /// <summary>
    /// <c>BaseAction.actionGroup</c>'s raw <c>[Flags]</c> BITMASK, the whole of
    /// it. <see cref="KspActionGroup"/> names the bits.
    ///
    /// <para>A mask rather than an ordinal because <c>KSPActionGroup</c> is a
    /// flags enum: one action can fire with several groups, which is exactly
    /// what <see cref="Groups"/> already carries as names. The mask is here
    /// because the NAME list cannot be trusted to be complete - it is built by
    /// intersecting the mask against the groups the capture knows about, so a
    /// group KSP adds is dropped before the wire and the client cannot tell that
    /// from a group nothing is bound to. The mask has no such ceiling.</para>
    /// </summary>
    [SitrepUnit(Units.Enumeration)]
    public int GroupsMask { get; set; }
}

/// <summary>
/// One resource row in <see cref="VesselPart.Resources"/>: storage
/// (<see cref="Amount"/>/<see cref="MaxAmount"/>) plus live flow
/// (<see cref="Flow"/>/<see cref="NominalFlow"/>).
///
/// <para><b>Flow scope.</b> <see cref="Flow"/>/<see cref="NominalFlow"/>
/// are populated only for the module types whose live rate is CHEAPLY
/// derivable from public fields without hand-simulating KSP's resource
/// solver: solar panels (<c>ModuleDeployableSolarPanel.flowRate</c>/
/// <c>chargeRate</c>), alternators (<c>ModuleAlternator.outputRate</c>),
/// and engine propellant consumption (<c>Propellant.currentRequirement</c>,
/// signed negative). This is the SAME "if cheap" scoping
/// <c>KspHost.BuildPartsPower</c>'s doc comment already establishes for
/// <c>totalProductionEc</c>: resource converters / fuel cells / drills
/// report storage only (no computed rate; not cheaply derivable from
/// static fields), matching that precedent rather than inventing a shaky
/// approximation. <see cref="NominalFlow"/> is omitted (left <c>null</c>)
/// whenever it would equal <see cref="Flow"/>, per the SDK contract.</para>
/// </summary>
[SitrepContract]
#if SITREP_CODEGEN
[TsInterface]
#endif
public class PartResourceFlow
{
    /// <summary><c>PartResource.amount</c>: current stored amount.</summary>
    [SitrepUnit(Units.ResourceUnits)]
    public double Amount { get; set; }

    /// <summary><c>PartResource.maxAmount</c>: storage capacity.</summary>
    [SitrepUnit(Units.ResourceUnits)]
    public double MaxAmount { get; set; }

    /// <summary>Signed units/sec: positive = producing, negative = consuming. <c>null</c> when no cheaply-derivable module contributes.</summary>
    [SitrepUnit(Units.ResourceUnitsPerSecond)]
    public double? Flow { get; set; }

    /// <summary>Same-sign 100%-efficiency cap (rated solar output). <c>null</c> when no module supports a nominal, or when it would equal <see cref="Flow"/>.</summary>
    [SitrepUnit(Units.ResourceUnitsPerSecond)]
    public double? NominalFlow { get; set; }
}

/// <summary>
/// One module's behavioural state in <see cref="VesselPart.ModuleStates"/>.
/// <see cref="Type"/> discriminates the module and <see cref="State"/> carries
/// the standardised deploy/activation word, whose vocabulary is on that
/// property.
/// </summary>
[SitrepContract]
#if SITREP_CODEGEN
[TsInterface]
#endif
public class PartModuleState
{
    /// <summary>Discriminator: <c>solarPanel</c> / <c>radiator</c> / <c>antenna</c> / <c>parachute</c> / <c>engine</c> / <c>drill</c> / <c>landingGear</c>. (<c>cargoBay</c> is a defined vocabulary value with no module here; see <c>KspHost.BuildPartModuleStates</c>'s doc comment for why.)</summary>
    [SitrepUnit(Units.Id)]
    public string Type { get; set; } = "";

    /// <summary>
    /// The standardised deploy/activation state, one closed vocabulary across
    /// every <see cref="Type"/> rather than each module's own enum spelling.
    ///
    /// <list type="bullet">
    /// <item><c>extended</c> / <c>retracted</c> / <c>deploying</c> /
    /// <c>retracting</c>, for anything that animates: solar panels, radiators,
    /// antennas, landing gear.</item>
    /// <item><c>stowed</c> / <c>armed</c> / <c>extended</c> / <c>broken</c>, the
    /// parachute lifecycle. <c>armed</c> is armed and waiting for its
    /// atmospheric trigger, which is not the same as deployed.</item>
    /// <item><c>active</c> / <c>inactive</c>, for engines and drills.</item>
    /// <item><c>unknown</c> when the underlying game enum maps to none of the
    /// above, which is a statement that the state was read and not recognised
    /// rather than a stand-in for retracted.</item>
    /// </list>
    /// </summary>
    [SitrepUnit(Units.Text)]
    public string State { get; set; } = "";

    /// <summary>Solar-panel-only: sun-tracking gimbal active. <c>null</c> for every other type.</summary>
    [SitrepUnit(Units.Flag)]
    public bool? Tracking { get; set; }

    /// <summary>Engine-only: fuel-starved. <c>null</c> unless <c>true</c> (never emitted as a bare <c>false</c>).</summary>
    [SitrepUnit(Units.Flag)]
    public bool? Flameout { get; set; }
}

/// <summary>
/// A <see cref="VesselPart"/>'s local bounding box: <see cref="Size"/> is the
/// part's <c>prefabSize</c> (a cheap, per-part-constant proxy for the renderer
/// bounds ShipMap could refine later), <see cref="Center"/> the mesh-centre
/// offset from the part's own origin (<c>Part.boundsCentroidOffset</c>).
/// Fuel-line parts report a whole-conduit-wrapping bounds, a carried-forward
/// KSP quirk the consumer handles, not this capture.
///
/// <para>Both are in the PART's own frame (<c>part-local</c>), never the
/// vessel's: they are authored fields of the part's config node, so one value
/// has to serve every instance of that part however it was assembled. A
/// consumer placing the box on a ship rotates both by that part's <c>orgRot</c>
/// before adding <see cref="VesselPart.Position"/>, which is <c>vessel-local</c>.
/// <internal>
/// <see cref="Center"/> was documented as vessel-local until the frame rule
/// landed; the config-node reading is what settled it.
/// </internal></para>
/// </summary>
[SitrepContract]
#if SITREP_CODEGEN
[TsInterface]
#endif
public class PartBounds
{
    /// <summary><c>Part.prefabSize</c>: the part's untransformed bounding-box extents (metres).</summary>
    [SitrepUnit(Units.Metres)]
    [SitrepFrame(Frames.PartLocal)]
    public Vec3 Size { get; set; } = new();

    /// <summary>
    /// <c>Part.boundsCentroidOffset</c>: mesh-centre offset from the part's own origin
    /// (metres, PART-local); <c>null</c> when absent.
    ///
    /// <para>Part-local, not vessel-local: an authored per-part-type constant like
    /// <see cref="Size"/> beside it, so one value serves every instance of the part at whatever
    /// attitude it was assembled at. <see cref="VesselPart.Position"/> is <c>vessel-local</c>,
    /// so adding the two raw mixes frames and lands the box somewhere the part is not; rotate
    /// this one by the part's <c>orgRot</c> first.</para>
    /// </summary>
    [SitrepUnit(Units.Metres)]
    [SitrepFrame(Frames.PartLocal)]
    public Vec3? Center { get; set; }
}

File: packages/core/src/schemas/vessel-parts.ts
/**
 * Vessel-topology and per-part state shapes. Originally part of the
 * Telemachus Reborn API schema; now also the shapes `@ksp-gonogo/data`'s
 * `useTopology` / `vesselPartsAdapter.ts` derive from the mod's
 * `vessel.parts` stream Topic.
 */

/**
 * One part in the `v.topology` payload. Shape mirrors the Telemachus
 * `PartsTopologyDataLinkHandler` walker exactly — no client-side renaming.
 * Live per-tick values (resources, temperatures) are NOT here; they live on
 * the separate `r.resourceFor[flightId]` / `therm.part[flightId]` keys so
 * the topology snapshot stays cacheable across staging-quiet stretches.
 */
export interface TopologyPart {
  flightId: number;
  persistentId: number;
  parentFlightId: number | null;
  /**
   * For parts carrying `CModuleFuelLine`, the flightId of the receiving
   * tank — the "to" end of the line. `parentFlightId` already points
   * at the source. `null` for any non-fuel-line part. Optional because
   * fixtures captured before the fork started emitting it stay readable.
   */
  fuelLineTarget?: number | null;
  name: string;
  title: string;
  manufacturer: string;
  /** `PartCategories` enum as string (e.g. "Engine", "FuelTank"). */
  category: string;
  inverseStage: number;
  crewCapacity: number;
  maxTemp: number;
  crashTolerance: number;
  /** `Part.mass` — dry mass only, not including resources. */
  dryMass: number;
  /** Vessel-local root-relative position `[x, y, z]`. */
  orgPos: [number, number, number];
  /**
   * Part's local +up axis in vessel-local frame (`part.orgRot * Vector3.up`).
   * `[0, 1, 0]` for axially-mounted parts (the majority); `[±1, 0, 0]` or
   * `[0, 0, ±1]` for radially-mounted parts (docking ports, side nose cones,
   * radial decouplers); inverted parts (rare) carry a negative-Y component.
   * Optional because fixtures captured before the fork started emitting it
   * stay readable — consumers should default to `[0, 1, 0]`.
   */
  up?: [number, number, number];
  /**
   * Prefab renderer bounds in metres — stable across the session.
   *
   * `size` is the mesh extent in part-local frame.
   *
   * `center` is the mesh-center offset from `orgPos` in *vessel-local*
   * frame (the fork pre-rotates by `orgRot` before emitting). Non-zero
   * for parts whose mesh doesn't sit on the attach-node anchor — radial
   * decouplers, surface ladders, structural brackets. Add it to `orgPos`
   * to get the mesh centre in assembly space. Optional because fixtures
   * captured before the fork started emitting it stay readable; missing
   * → treat as `{x: 0, y: 0, z: 0}` (correct for the axially-stacked
   * majority where the mesh centres on the anchor).
   */
  bounds: {
    size: { x: number; y: number; z: number };
    center?: { x: number; y: number; z: number };
  };
  /** Raw `PartModule.moduleName` strings; no filtering. */
  modules: string[];
}

/**
 * `v.topology` response shape. `topologySeq` matches the lightweight
 * `v.topologySeq` key — consumers subscribe to the seq and refetch this key
 * only when it ticks rather than streaming the topology continuously.
 */
export interface VesselTopology {
  topologySeq: number;
  rootFlightId: number;
  parts: TopologyPart[];
}

/**
 * Live per-part resource state. Keyed by resource name (LiquidFuel,
 * Oxidizer, ElectricCharge, ...); empty object when the part has no
 * resources / contributes no flow / the flightId isn't found. Sourced from
 * the mod's `vessel.parts` stream Topic (`@ksp-gonogo/data`'s
 * `usePartsLive`/`derivePartResources`) — originally the legacy fork's
 * `r.resourceFor[flightId]` key, whose row shape this interface still
 * mirrors field-for-field.
 *
 * `amount` / `maxAmount` cover storage. `flow` and `nominalFlow` are
 * present when the part's modules contribute production / consumption
 * (solar panels, RTGs, generators, ISRU, drills, engines):
 *
 * - `flow`: signed units/sec — positive = producing, negative =
 *   consuming. Summed across every contributing module on the part.
 * - `nominalFlow`: same sign, 100%-efficiency cap (sun-aligned solar,
 *   full-rate generator). Omitted when no module supports a nominal
 *   (e.g. engines), when no module contributes, or when nominal equals
 *   flow.
 *
 * Rows are emitted for resources the part contributes flow to even when
 * storage is zero — an RTG part reports `{ ElectricCharge: { amount: 0,
 * maxAmount: 0, flow: 0.7, nominalFlow: 0.75 } }`.
 */
export interface PartResources {
  [resourceName: string]: {
    amount: number;
    maxAmount: number;
    flow?: number;
    nominalFlow?: number;
  };
}

/**
 * Live per-part thermal state from `therm.part[flightId]`. `null` upstream
 * when the flightId isn't found — consumers should treat the missing case
 * as "thermal data not available" rather than zero-Kelvin.
 */
export interface PartThermal {
  temperature: number;
  maxTemperature: number;
  temperatureK: number;
  maxTemperatureK: number;
}

/**
 * Behavioural state for one PartModule on a part — the fork emits one
 * entry per supported module under `v.partState[flightId].modules`.
 *
 * `type` discriminates the module; `state` is the standardised deploy /
 * activation vocabulary (see fork's PartStateDataLinkHandler):
 *
 * - `extended` / `retracted` / `deploying` / `retracting` — for parts
 *   that animate (solar panels, radiators, antennas, cargo bays, gear).
 * - `stowed` / `armed` / `extended` / `broken` — parachute lifecycle.
 *   `armed` means the chute is armed and waiting for atmospheric trigger.
 * - `active` / `inactive` — engines (EngineIgnited toggle), drills.
 * - `unknown` — fallback when the underlying KSP enum doesn't map.
 *
 * Module-specific extras:
 * - solarPanel: `tracking` (bool) — sun-tracking gimbal active.
 * - engine: `flameout` (bool, only when true) — fuel-starved.
 */
export interface PartStateModule {
  type:
    | "solarPanel"
    | "radiator"
    | "antenna"
    | "parachute"
    | "engine"
    | "drill"
    | "cargoBay"
    | "landingGear";
  state: string;
  tracking?: boolean;
  flameout?: boolean;
}

/**
 * Per-part module behavioural state. Originally the legacy fork's
 * `v.partState[flightId]` response, where `seq` was a per-vessel counter
 * bumped on staging, vessel modifications, part death / undock / couple,
 * PAW dismiss, and as a 10s backstop for events without a GameEvent hook —
 * consumers dedup on `seq` rather than deep-comparing the modules array.
 *
 * Now sourced from the mod's `vessel.parts` stream Topic
 * (`@ksp-gonogo/data`'s `usePartsLive`/`derivePartState`), whose payload
 * re-emits atomically on any change — there's no separate per-part dedup
 * counter on the wire any more, so `seq` is synthesized from the module
 * count (see `derivePartState`'s doc comment). No current consumer reads
 * `.seq` directly.
 */
export interface PartState {
  seq: number;
  modules: PartStateModule[];
}
